src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+                self.wal_path.display()
+            );
+        }
+
+        Ok(segments.pop().unwrap())
+    }
+
+    fn find_batches(&self) -> Result<Vec<Batch>, anyhow::Error> {
+        lazy_static! {
+            static ref BATCH_REGEX: Regex = Regex::new("^batch-[0-9]+-[0-9]+$").unwrap();
+        }
+
+        let batches = read_dir_regex(&self.trace_path, &BATCH_REGEX)?;
+
+        let mut batches: Vec<Batch> = batches
+            .into_iter()
+            .map(Batch::reinit)
+            .collect::<Result<_, _>>()
+            .unwrap();
+        batches.sort_by_key(|batch| batch.lower);
+        Ok(batches)
+    }
+
+    /// Try to compact all of the batches we know about into a single batch from
+    /// [compaction_frontier, upper)
+    ///
+    /// TODO: the approach to compacting is likely very suboptimal.
+    fn compact(&mut self) -> Result<(), anyhow::Error> {
+        self.consume_wal()?;
+
+        if self.batches.len() > 10 {
+            let batches = std::mem::replace(&mut self.batches, vec![]);
+            let batch = Batch::compact(&batches, &self.trace_path, self.compaction)?;
+            self.batches.push(batch);
+
+            // TODO: This seems like potentially a place with a weird failure mode.
+            for batch in batches {
+                fs::remove_file(&batch.path).with_context(|| {
+                    format!("failed to remove replaced batch {}", batch.path.display())
+                })?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Re-initialize a trace based on the available batch files on disk.
+    pub fn resume(
+        id: GlobalId,
+        traces_path: &Path,
+        wals_path: &Path,
+    ) -> Result<Self, anyhow::Error> {
+        // Need to instantiate a new trace and figure out what batches
+        // we have access to.
+        let trace_path = traces_path.join(id.to_string());
+        let wal_path = wals_path.join(id.to_string());
+        let mut ret = Self {
+            trace_path,
+            wal_path,
+            batches: Vec::new(),
+            compaction: None,
+        };
+
+        let mut batches = ret.find_batches()?;
+        ret.batches.append(&mut batches);
+
+        Ok(ret)
+    }
+
+    /// Read in the data for this relation, from all available batches and
+    /// WAL segments back into memory.
+    pub fn read(&self) -> Result<Vec<Message>, anyhow::Error> {
+        let mut out = vec![];
+
+        for batch in self.batches.iter() {
+            let mut messages = batch.read()?;
+            out.append(&mut messages);
+        }
+
+        let finished_segments = self.find_finished_wal_segments()?;
+        let unfinished_segment = self.find_unfinished_wal_segment()?;
+
+        for segment in finished_segments {
+            let mut messages = read_segment(&segment)?;
+            out.append(&mut messages);
+        }
+
+        out.append(&mut read_segment(&unfinished_segment)?);
+
+        out.dedup();
```

Comment:
Is this dedup for the potentially duplicated frontier messages or something else? It's unclear to me if `out` needs to be sorted first